caddyconfig/httpcaddyfile/httptype.go

```diff
@@ -43,6 +43,16 @@ func (st ServerType) Setup(originalServerBlocks []caddyfile.ServerBlock,
 	var warnings []caddyconfig.Warning
 	gc := counter{new(int)}
 
+	state := make(map[string]interface{})
+	// tlsCertTags maps certificate filenames to their tag.
+	// This is used to remember which tag is used for each
+	// certificate files, since we need to avoid loading
+	// the same certificate files more than once, overwriting
+	// previous tags
+	state["tlsCertTags"] = make(map[string]string)
+	// this tracks the logCounter across all servers
+	state["logCounter"] = 0
```

Comment:
So, the reason I want to avoid initializing up here in Setup() is because the Caddyfile adapter is extensible, and I don't want to put plugins' logic into the core of the adapter. Otherwise we'll have to keep changing the core to support all the plugins.

Instead, let's move all this initialization to where we first use them, like so:

```go
tlsCertTags, ok := h.state["tlsCertTags"].(map[string]string)
if !ok {
    tlsCertTags = make(map[string]string)
    h.state["tlsCertTags"] = tlsCertTags
}
// then proceed to use it
```

Does that make sense?

Reply:
I see but leave state where is it, outside of all of the server blocks.

Reply:
Yes, exactly -- ignore my wrong comment below. :smile:

Reply:
Done